Font Library: fix duplicate variants with different file types

jffng opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This PR adds checks to ensure that multiple file types of the same font variant are not added, so that only one variation per font style and weight is stored. 

## Why?
Fixes #54253

## How?
For fonts where the font family does not yet exist, checks the list of new font faces and compares their font weight and style, and if there is already an existing font face, skips adding the incoming one.

For font faces where the font family exists, adds a new function `get_intersecting_font_faces` which returns an array of font faces to be deleted and replaced by the incoming font face.

## Testing Instructions
1. Simultaneously (batch) upload two or more font faces of the same family, style, and weight, but different file types. 
2. Ensure only the first variant is added / used.
3. Upload a font face to an existing font family with the same style and weight, but different file type.
4. Ensure the uploaded version is used / replaces the existing one, and an additional variant is not created.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

jffng committed:
Add get_intersecting_font_faces to determine duplicates to be removed.

jffng committed:
Add duplicant font variant test.

jffng committed:
Make linter happy.

jffng committed:
Use local fonts for duplicate font variant test.

jffng committed:
Make linter happy.

pbking reviewed (approved):
This is working as expected for me.  Fonts are limited to a single format.  When adding a new it seems to always replace the existing, whatever format it is.  

When uploading multiple at the same time the format chosen seems to be... alphabetical? 

Either way, I can confirm that it fixes the issue at hand.

jffng commented on lib/experimental/fonts/font-library/class-wp-font-family.php:
```diff
@@ -493,7 +530,21 @@ private function create_font_post() {
 		$post_font_data = json_decode( $post->post_content, true );
 		$new_data       = WP_Font_Family_Utils::merge_fonts_data( $post_font_data, $this->data );
-		$this->data     = $new_data;
+		$intersecting   = $this->get_intersecting_font_faces( $post_font_data['fontFace'], $new_data['fontFace'] );
```
Can we expect that the data will always contain a `fontFace` key, or should we be more defensive here?

matiasbenedetto replied:
Good catch, nope, there can be font families without a fontFace array defined or an empty array. So a check for that seems necessary,

jffng replied:
Thanks, done in [b07fdd0](https://github.com/WordPress/gutenberg/pull/54490/commits/b07fdd083687c250a396588cb9b525d6c3cb568c)

matiasbenedetto commented:
The tests I made manually are working great.
Currently, we are testing the case when two font faces have the same characteristics but different font files in the same request.
Could we add one more unit test case to cover the addition of font faces with the same characteristics in different requests? That's covered in the implementation but not in the unit tests.

jffng committed:
Ensure fontFace exists and is not empty.

jffng committed:
WIP test.

matiasbenedetto commented on phpunit/tests/fonts/font-library/wpRestFontLibraryController/installFonts.php:
```diff
@@ -414,4 +414,91 @@ public function data_install_with_improper_inputs() {
+	 * @param array $font_families Font families to install in theme.json format.
+	 * @param array $files         Font files to install.
+	 */
+	public function test_install_with_duplicate_font_faces( $font_families, $files = array(), $expected_response ) {
+		$install_request    = new WP_REST_Request( 'POST', '/wp/v2/fonts' );
+		$font_families_json = json_encode( $font_families );
+		$install_request->set_param( 'fontFamilies', $font_families_json );
+		$install_request->set_file_params( $files );
+
+		$response = rest_get_server()->dispatch( $install_request );
+		$this->assertSame( 200, $response->get_status(), 'The response status is not 200.' );
+		$data = $response->get_data();
+		$this->assertCount( 1, $expected_response[0]['fontFace'], 'Duplicate font faces were not removed / added incorrectly.' );
+		$this->assertStringEndsWith( $expected_response[0]['fontFace'][0]['src'], $data[0]['fontFace'][0]['src'], 'The src attribute does not match the expected font file.' );
```
As the de-duplication functionality is not in the Rest controller but in the Font Family class, it may be better to move this test to the install tests file: `phpunit/tests/fonts/font-library/wpFontFamily/install.php`.

Apart from that, as we are not testing the API endpoint, I think the requests are not necessary because we are not testing that. Probably it would be better to add the data we want to test like this:


```
$font = new WP_Font_Family( $the_font_data_im_testing )
$font->install();
```

See this test as an example: https://github.com/WordPress/gutenberg/blob/761c390e3ce0a92a9f4183509074d634799711ef/phpunit/tests/fonts/font-library/wpFontFamily/install.php#L86-L101

jffng replied:
Good call, done in [ce5a4a9](https://github.com/WordPress/gutenberg/pull/54490/commits/ce5a4a9cd393ebed176c6bf2df9a72be86c1daf5).

jffng committed:
Move tests to fontFamily class and add another test when installing more than one font face.

jffng committed:
Add test that only a single font face extension is maintained, overwriting initial.

jffng committed:
Format php.

jffng commented:
@matiasbenedetto I moved these tests to the `Tests_Fonts_WpFontFamily_Install` class, added a case to cover installing multiple font faces that contain duplicates, and one test to cover the addition of font faces with the same characteristics in different requests.

jeffikus reviewed (approved):
LGTM - I tested with with Cardo font.

- Regular 
- Italic
- Bold

Used ttf, woff, woff2 file types for each.

This created 3 variants within Cardo, tried uploading 9 files total.
